File: pkg/vision/color.go
package vision

import "math"

// RGBToHSV 把 8-bit RGB 转 HSV。返回 H ∈ [0,360]，S/V ∈ [0,100]。
//
// 跟主流取色器 (Photoshop/Windows/Figma) 对齐：H 满程角度、S/V 百分比，用户截图取色
// 读到的值能直接填进 DetectColor 的 hsv 阈值。统一一份实现避免漂移：H 算法若漏 60 倍
// 精度，会让阈值对不上取色工具调出的值。
func RGBToHSV(r, g, b uint8) (h int, s int, v int) {
	rf := float64(r) / 255.0
	gf := float64(g) / 255.0
	bf := float64(b) / 255.0
	maxC := math.Max(rf, math.Max(gf, bf))
	minC := math.Min(rf, math.Min(gf, bf))
	delta := maxC - minC

	v = int(maxC * 100)
	if maxC == 0 {
		return 0, 0, v
	}
	s = int((delta / maxC) * 100)

	var hf float64
	switch {
	case delta == 0:
		hf = 0
	case maxC == rf:
		hf = 60 * math.Mod((gf-bf)/delta, 6)
	case maxC == gf:
		hf = 60 * ((bf-rf)/delta + 2)
	default:
		hf = 60 * ((rf-gf)/delta + 4)
	}
	if hf < 0 {
		hf += 360
	}
	h = int(hf)
	return
}

File: frontend/src/lib/color.ts
// RGB↔HSV — 前端版, 跟 Go pkg/vision.RGBToHSV 同一套约定: H ∈ [0,360], S/V ∈ [0,100].
// Screenshot color-pick values use the same scale as Workflow color analysis.
// 改这里前先看 pkg/vision/color.go, 别让前后端漂移.

export interface HSV {
  h: number
  s: number
  v: number
}

// rgbToHsv 8-bit RGB (0-255) → HSV (H 0-360, S/V 0-100). 取整跟 Go 侧一致 (截断, 非四舍五入).
export function rgbToHsv(r: number, g: number, b: number): HSV {
  const rf = r / 255
  const gf = g / 255
  const bf = b / 255
  const maxC = Math.max(rf, gf, bf)
  const minC = Math.min(rf, gf, bf)
  const delta = maxC - minC

  const v = Math.trunc(maxC * 100)
  if (maxC === 0) return { h: 0, s: 0, v }
  const s = Math.trunc((delta / maxC) * 100)

  let hf: number
  if (delta === 0) hf = 0
  else if (maxC === rf) hf = 60 * (((gf - bf) / delta) % 6)
  else if (maxC === gf) hf = 60 * ((bf - rf) / delta + 2)
  else hf = 60 * ((rf - gf) / delta + 4)
  if (hf < 0) hf += 360

  return { h: Math.trunc(hf), s, v }
}

// rgbToHex 8-bit RGB → '#RRGGBB' (大写, 跟 Go 侧 fmt '%02X' 一致).
export function rgbToHex(r: number, g: number, b: number): string {
  const hx = (n: number) => n.toString(16).padStart(2, '0').toUpperCase()
  return `#${hx(r)}${hx(g)}${hx(b)}`
}
